### DISCUSSION TITLE
Fallback to receive max_identfier_length for Oracle

### DISCUSSION THREAD

**Discussion by finswimmer** [sqlalchemy/sqlalchemy#11879] [Usage Questions]
Hello,

according to the [docs](https://docs.sqlalchemy.org/en/20/dialects/oracle.html#max-identifier-lengths) sqlalchemy is trying to fetch the "compatible" version to determine the max_identifier_length. The problem with the query used is, that `v$parameter` cannot be read by "normal" users usually and (according to our dba's) it's also not recommended by Oracle to grant access to it.

Would it be possible to implement another fallback, which is doing an educated guess to get to allowed length. Something like trying `SELECT 'foo' as <128_char_long_string> FROM DUAL;`, `SELECT 'foo' as <30_char_long_string> FROM DUAL;` and check what works?

Thanks a lot,

fin swimmer

**Comment by CaselIT** [comment]
Hi,

I'm not sure why we use that query to verify the max identifier length.
Did your dba suggest a particular query to obtain it, or was the suggestion to use a `select ...`?

@zzzeek can you think of a reason why trying selecting a long label would not work in this case?

**Comment by zzzeek** [comment]
What's the actual problem we are trying to solve here?  your DBAs don't like seeing that query in the logs ?

**Reply by zzzeek** [reply]
basically I dont want to change anything here, I want a discussion posted at https://github.com/oracle/python-oracledb where we will ask them how they have [connection.version](https://python-oracledb.readthedocs.io/en/latest/api_manual/connection.html#Connection.version) to provide the server version, however this is incomplete since "compatibility" indicates the server will act like a different version.     Whatever recommendation @anthony-tuininga gives us, that's what we'd go with

**Reply by finswimmer** [reply]
Hey @zzzeek,

> What's the actual problem we are trying to solve here? 

the problem we are facing is this: We are using oracle 19c with a compatible version set to 11. This means the database doesn't accept identifiers with more than 30 characters. On connect sqlalchemy recognize that version 19 is used and sets `max_identfier_length` to 128 as default. Because `v$parameter` cannot be read, it cannot recognize that only 30 characters are allowed. So sqlalchemy will generate for example `ALIAS` names with more than 30 characters, which will be rejected by the database.

Our dba's hesitate to grant access for users to `v$parameter` because:
* oracle doesn't recommend it
* it has to be done for every user

fin swimmer

**Reply by zzzeek** [reply]
oh OK you actually have the exact problem that you're using that parameter and the DBAs dont want to grant access.      OK can we ask on the oracledb discussion tracker what their best practice here is, if they say to "select a 128 char ident and (catch the error?  does it raise" then we'll go with that.

you can work around the whole thing by just sending the identifier length you want:

     e = create_engine("oracle+oracledb://...", max_identifier_length=30)

**Reply by zzzeek** [reply]
https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.max_identifier_length

**Reply by anthony-tuininga** [reply]
@zzzeek, I'll ask internally. I can see the value is available easily enough in thin mode and I could make it available as a connection property. In thick mode I don't see any real way of doing so currently but I'll ask and see if there is and get back to you. Please do log a discussion at oracledb and we'll get back to you.

**Reply by CaselIT** [reply]
@anthony-tuininga I can open a feature request at oracledb. 
> I can see the value is available easily enough

here are you talking about the max identifier length, or the compat version returned by `SELECT value FROM v$parameter WHERE name = 'compatible'`? Either one would work as property for sqlalchemy needs.

**Reply by anthony-tuininga** [reply]
> here are you talking about the max identifier length, or the compat version returned by SELECT value FROM v$parameter WHERE name = 'compatible'? Either one would work as property for sqlalchemy needs.

I am referring to the max identifier length.

**Reply by CaselIT** [reply]
opened https://github.com/oracle/python-oracledb/issues/395 in oracledb to track it.

Once that's available the oracledb dialect can use that property instead of relying on a query

**Reply by zzzeek** [reply]
is there any way we can get the effective server version?   im not sure how helpful `connection.version` is if the database is placed into a special mode where it behaves like some other arbitrary version.   dont the Oracle developers feel this is something the client might need to know?

**Reply by anthony-tuininga** [reply]
Understood. I'll ask!

**Reply by CaselIT** [reply]
For this use case either property would work for sqlalchemy, since the compatible version is only used to decide the max length, but if we would have to choose one it's probably more useful to have the compatible version available

**Reply by anthony-tuininga** [reply]
The changes in oracledb are complete and will be made available in version 2.5 when that releases (hopefully fairly soon). With that in place there is no need to query the compatible value since the `connection.max_identifier_length` is accurately returned when compatible is set to a lower version than the actual database version.

**Reply by CaselIT** [reply]
I've added https://github.com/sqlalchemy/sqlalchemy/issues/12032 to add support to this new attribute!
